File: OsmAnd-java/src/main/java/net/osmand/binary/QueryToken.java
package net.osmand.binary;

import gnu.trove.list.array.TIntArrayList;
import net.osmand.Collator;
import net.osmand.CollatorStringMatcher;

import java.util.*;

public class QueryToken {
    final String query;
    final List<Prefix> prefixes;
    final Collator collator;
    final CollatorStringMatcher.StringMatcherMode matcherMode;
    
    record Prefix(String key, int offset) {}

    class SuffixMask {
        TIntArrayList masks;
        final Prefix prefix;
        private boolean passThrough;
        int prevMask = 0; // not exactly correct to maintain state here

        SuffixMask(Prefix prefix) {
            this.prefix = prefix;
        }

        void setDictionary(List<String> suffixDictionary) {
            passThrough = suffixDictionary == null;
            if (prefix.key() == null || suffixDictionary == null) {
                return;
            }
            
            if (suffixDictionary.size() == 1 && suffixDictionary.get(0).isEmpty()) {
                passThrough = query != null && CollatorStringMatcher.cmatches(collator, prefix.key(), query, matcherMode);
                return;
            }
            if (masks == null) {
                masks = new TIntArrayList();
            }
            if (query == null) {
                return;
            }
            for (int index = 0; index < suffixDictionary.size(); index++) {
                addSuffix(index, suffixDictionary.get(index));
            }
        }

        boolean shouldPassThrough() {
            return passThrough;
        }
        
        boolean isMatched(int maskIndex, int mask) {
			if (masks == null) {
				return true;
			}
			if (maskIndex == 0) {
				prevMask = 0;
			}
			boolean res = false;
			// use only masks for first and after delimiter
			if (prevMask == 0 && mask % 2 == 0 && masks.contains(mask / 2 - 1)) {
				res = true;
			}
			prevMask = mask;
			return res;
        }

        private void addSuffix(int index, String suffix) {
            if (suffix == null || index < 0) {
                return;
            }
            String fullKey = prefix.key() + suffix;
            if (CollatorStringMatcher.cmatches(collator, fullKey, query, matcherMode)) {
            	masks.add(index);
            }
        }
    }

    QueryToken(String query, Collator collator, CollatorStringMatcher.StringMatcherMode matcherMode, List<Prefix> prefixes) {
        this.query = query;
        this.collator = collator;
        this.matcherMode = matcherMode;

        if (prefixes == null || prefixes.isEmpty()) {
            this.prefixes = Collections.emptyList();
        } else {
            this.prefixes = new ArrayList<>(prefixes);
            this.prefixes.sort((left, right) -> {
                int lengthCompare = Integer.compare(right.key.length(), left.key().length());
                if (lengthCompare != 0) {
                    return lengthCompare;
                }
                return left.key().compareTo(right.key());
            });
        }
    }

	public boolean matchFullPrefix(String key) {
		return CollatorStringMatcher.cmatches(collator, key, query, matcherMode);
	}
	
}

File: OsmAnd-shared/src/commonMain/kotlin/net/osmand/shared/binary/QueryToken.kt
package net.osmand.shared.binary

import net.osmand.shared.api.KStringMatcherMode
import net.osmand.shared.util.KCollationKey
import net.osmand.shared.util.KCollatorStringMatcher.Companion.cmatchesKey
import net.osmand.shared.util.collections.KTIntArrayList

/**
 * Copy of `net.osmand.binary.QueryToken` in OsmAnd-java: one word of the query, the prefixes of the
 * poi name index it could sit under, and which of that prefix's suffixes it can still match.
 *
 * The query is reduced to a [KCollationKey] once here, where java rebuilds the comparison from the
 * string on every suffix of every prefix.
 *
 * Internal like the java original, whose constructor is package private: it is scaffolding for
 * reading the name index, not something a caller outside the reader builds.
 */
internal class QueryToken(
	val query: String?,
	val matcherMode: KStringMatcherMode,
	prefixes: List<Prefix>?
) {

	private val queryKey: KCollationKey? = query?.let { KCollationKey.of(it) }

	/** Longest first, then alphabetically, so that the most specific prefix is tried first. */
	val prefixes: List<Prefix> = if (prefixes.isNullOrEmpty()) {
		emptyList()
	} else {
		prefixes.sortedWith(compareByDescending<Prefix> { it.key.length }.thenBy { it.key })
	}

	data class Prefix(val key: String, val offset: Int)

	fun matchFullPrefix(key: String): Boolean = matches(key)

	private fun matches(fullName: String): Boolean {
		val part = queryKey ?: return false
		return cmatchesKey(fullName, part, alignFull = true, mode = matcherMode)
	}

	inner class SuffixMask(private val prefix: Prefix) {

		private var masks: KTIntArrayList? = null
		private var passThrough: Boolean = false

		/** Not exactly correct to maintain state here, as the java original says. */
		private var prevMask: Int = 0

		fun setDictionary(suffixDictionary: List<String>?) {
			passThrough = suffixDictionary == null
			if (suffixDictionary == null) {
				return
			}
			if (suffixDictionary.size == 1 && suffixDictionary[0].isEmpty()) {
				passThrough = query != null && matches(prefix.key)
				return
			}
			if (masks == null) {
				masks = KTIntArrayList()
			}
			if (query == null) {
				return
			}
			for (index in suffixDictionary.indices) {
				addSuffix(index, suffixDictionary[index])
			}
		}

		fun shouldPassThrough(): Boolean = passThrough

		fun isMatched(maskIndex: Int, mask: Int): Boolean {
			val masks = this.masks ?: return true
			if (maskIndex == 0) {
				prevMask = 0
			}
			var res = false
			// use only masks for first and after delimiter
			if (prevMask == 0 && mask % 2 == 0 && masks.contains(mask / 2 - 1)) {
				res = true
			}
			prevMask = mask
			return res
		}

		private fun addSuffix(index: Int, suffix: String?) {
			if (suffix == null || index < 0) {
				return
			}
			if (matches(prefix.key + suffix)) {
				masks?.add(index)
			}
		}
	}
}
